File: tools/place_layout_skp.py
"""place_layout_skp.py — materializa o layout VENCEDOR (Etapa B+C) no .skp real
como PLACEHOLDERS (boxes coloridos extrudados), pra o Felipe ver o layout na
planta_74 e dar o veredito visual. NAO usa 3D Warehouse, NAO baixa asset.

Slice minimo (gate :8765, opcao A): so a sala r002, m013 marcado AMBIGUOUS,
render before/after pra VISUAL_REVIEW. Felipe 2026-06-04.

Coordenadas: usa a MESMA conversao do furnish_plan (SU = pdf * PT_TO_IN, sem
flip de Y) que comprovadamente bate com o shell artifacts/planta_74.skp.

Uso: python tools/place_layout_skp.py [--room r002]
"""
import argparse
import json
import os
import subprocess
import sys
import time
from pathlib import Path

from tools.layout_candidates import EXTRA_TEMPLATES, TEMPLATES, _tv_setup, run

ROOT = Path(r"E:\Claude\sketchup-mcp")
SKETCHUP_EXE = r"C:\Program Files\SketchUp\SketchUp 2026\SketchUp\SketchUp.exe"
PT_TO_IN = (0.19 / 5.4) * 39.3700787402
CONSENSUS = ROOT / "fixtures/planta_74/consensus_with_human_walls_and_soft_barriers.json"
BASE_SKP = ROOT / "artifacts/planta_74/planta_74.skp"
OUT_DIR = ROOT / "runs/planta_74"

# altura de extrusao do placeholder por tipo (m) e cor RGB (= cores do PNG 2D)
KIND_H_M = {"sofa_3": 0.80, "sofa_2": 0.80, "rack_tv": 0.45,
            "mesa_centro": 0.40, "poltrona": 0.80,
            "tapete": 0.02, "aparador": 0.85}
KIND_RGB = {"sofa_3": [21, 101, 192], "sofa_2": [21, 101, 192],
            "rack_tv": [106, 27, 154], "mesa_centro": [239, 108, 0],
            "poltrona": [0, 131, 143], "tapete": [201, 185, 160],
            "aparador": [121, 85, 72]}


def build_boxes(con, room_id, template_name=None):
    """Roda o cerebro de layout e converte o layout em boxes SU (inches).
    Sem template_name: materializa o VENCEDOR do ranking. Com template_name:
    materializa esse template direto (offset 0), util pra prototipar composicao
    antes do score premia-la."""
    sm, out = run(con, room_id)
    tv = _tv_setup(sm)
    if tv is None:
        return None, out
    ambiguous = out["tv_wall"].get("confidence") == "ambiguous"
    if template_name:
        fn = EXTRA_TEMPLATES.get(template_name) or dict(TEMPLATES).get(template_name)
        if fn is None:
            return None, {"result": "NO_TEMPLATE",
                          "reason": f"template '{template_name}' desconhecido",
                          "tv_wall": out["tv_wall"]}
        items_raw = fn(tv)
        out["chosen_candidate"] = template_name
    elif out["result"] != "OK":
        return None, out
    else:
        cand = next(c for c in out["candidates"] if c["template"] == out["chosen_candidate"])
        items_raw = cand["_items"]
    boxes = []
    for it in items_raw:
        x0, y0, x1, y1 = it["box"].bounds
        kind = it["kind"]
        is_tv_wall = kind == "rack_tv"   # rack mora na parede-TV
        # cantos reais do poligono (preserva rotacao p/ moveis angulados); o .rb
        # desenha via add_face(corners), nao via bbox.
        coords = list(it["box"].exterior.coords)[:-1]
        corners = [[round(px * PT_TO_IN, 2), round(py * PT_TO_IN, 2)] for px, py in coords]
        boxes.append({
            "kind": kind,
            "x0": x0 * PT_TO_IN, "y0": y0 * PT_TO_IN,
            "x1": x1 * PT_TO_IN, "y1": y1 * PT_TO_IN,
            "corners": corners,
            "h_in": KIND_H_M.get(kind, 0.5) * 39.3700787402,
            "rgb": KIND_RGB.get(kind, [120, 120, 120]),
            "label": kind + ("_AMBIGUOUS" if (is_tv_wall and ambiguous) else ""),
            "ambiguous": bool(is_tv_wall and ambiguous),
            "decorative": bool(it.get("decorative")),
        })
    return boxes, out


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("--room", default="r002")
    ap.add_argument("--template", default=None,
                    help="forcar um template (ex: estar_ancorado); default = vencedor do score")
    ap.add_argument("--dry-run", action="store_true",
                    help="monta os boxes e imprime, sem lancar o SketchUp")
    args = ap.parse_args()
    con = json.loads(CONSENSUS.read_text("utf-8"))
    boxes, out = build_boxes(con, args.room, args.template)
    if boxes is None:
        print(f"[place] {out['result']}: {out.get('reason')} — nada a materializar")
        sys.exit(2)

    if args.dry_run:
        print(f"[dry-run] sala {args.room} | vencedor {out['chosen_candidate']} | "
              f"{len(boxes)} boxes (SU inches):")
        for b in boxes:
            print(f"  {b['label']:22} x[{b['x0']:.0f}..{b['x1']:.0f}] "
                  f"y[{b['y0']:.0f}..{b['y1']:.0f}] h={b['h_in']:.0f} rgb={b['rgb']}")
        if out.get("tv_wall_uncertainty"):
            print(f"  /!\\ {out['tv_wall_uncertainty']}")
        return

    chosen = out["chosen_candidate"]
    score = next((c["total_score"] for c in out.get("candidates", []) if c["template"] == chosen), None)
    score_str = f"{score} pts" if score is not None else "(template forcado, fora do ranking)"
    print(f"[place] sala {args.room} | layout: {chosen} {score_str} | {len(boxes)} placeholders")
    if out.get("tv_wall_uncertainty"):
        print(f"[place] /!\\ {out['tv_wall_uncertainty']}")

    OUT_DIR.mkdir(parents=True, exist_ok=True)
    tag = f"_{args.template}" if args.template else ""
    skp_out = OUT_DIR / f"planta_74_layout_{args.room}{tag}.skp"
    before_png = OUT_DIR / f"layout_{args.room}{tag}_before_top.png"
    after_top = OUT_DIR / f"layout_{args.room}{tag}_after_top.png"
    after_iso = OUT_DIR / f"layout_{args.room}{tag}_after_iso.png"
    log_path = OUT_DIR / f"layout_{args.room}{tag}_place_log.txt"
    for p in (skp_out, before_png, after_top, after_iso, log_path):
        if p.exists():
            p.unlink()

    env = os.environ.copy()
    env["LAYOUT_BOXES"] = json.dumps(boxes)
    env["LAYOUT_OUT"] = str(skp_out).replace("\\", "/")
    env["LAYOUT_BEFORE"] = str(before_png).replace("\\", "/")
    env["LAYOUT_AFTER_TOP"] = str(after_top).replace("\\", "/")
    env["LAYOUT_AFTER_ISO"] = str(after_iso).replace("\\", "/")
    env["LAYOUT_LOG"] = str(log_path).replace("\\", "/")

    subprocess.run(["taskkill", "/F", "/IM", "SketchUp.exe"], capture_output=True)
    time.sleep(1)
    cmd = [SKETCHUP_EXE, str(BASE_SKP), "-RubyStartup", str(ROOT / "tools/place_layout_skp.rb")]
    print(f"[place] launching SU (base={BASE_SKP.name})...")
    subprocess.Popen(cmd, env=env, creationflags=getattr(subprocess, "DETACHED_PROCESS", 0))

    deadline = time.time() + 240
    while time.time() < deadline:
        if log_path.exists():
            time.sleep(2)
            break
        time.sleep(1)
    subprocess.run(["taskkill", "/F", "/IM", "SketchUp.exe"], capture_output=True)

    if log_path.exists():
        print("[place] LOG:")
        print(log_path.read_text("utf-8"))
    else:
        print("[place] TIMEOUT — SU nao produziu log")
        sys.exit(1)

    # cross-check deterministico de coordenadas (o gate alertou: offset/escala)
    print("\n[place] === COORD CROSS-CHECK ===")
    bx = [b["x0"] for b in boxes] + [b["x1"] for b in boxes]
    by = [b["y0"] for b in boxes] + [b["y1"] for b in boxes]
    print(f"  bbox moveis (SU in): x[{min(bx):.0f}..{max(bx):.0f}] "
          f"y[{min(by):.0f}..{max(by):.0f}]")
    print(f"  (se no render TOP os boxes caem DENTRO da sala da planta, coord OK)")
    print(f"\n[place] artefatos -> {OUT_DIR}/")
    print(f"  SKP:    {skp_out.name}")
    print(f"  BEFORE: {before_png.name}")
    print(f"  AFTER:  {after_top.name} / {after_iso.name}")


if __name__ == "__main__":
    main()

File: tools/place_layout_skp.rb
# place_layout_skp.rb — desenha o layout VENCEDOR como PLACEHOLDERS (boxes
# coloridos extrudados) no shell da planta_74 ja aberto, renderiza before/after.
# Rodado via:  SketchUp.exe <base.skp> -RubyStartup place_layout_skp.rb
# Params via ENV:
#   LAYOUT_BOXES = JSON [{kind,x0,y0,x1,y1 (SU in),h_in,rgb:[r,g,b],label,ambiguous}]
#   LAYOUT_OUT / LAYOUT_BEFORE / LAYOUT_AFTER_TOP / LAYOUT_AFTER_ISO / LAYOUT_LOG
# Placeholders, NAO 3D Warehouse. Felipe 2026-06-04 (gate :8765 opcao A).
require 'json'

def pl_top_camera(model, bbox = nil)
  view = model.active_view
  bbox ||= model.bounds
  center = bbox.center
  eye = Geom::Point3d.new(center.x, center.y, center.z + bbox.diagonal * 5.0)
  cam = Sketchup::Camera.new(eye, center, Geom::Vector3d.new(0, 1, 0))
  cam.perspective = false
  mw = bbox.max.x - bbox.min.x
  mh = bbox.max.y - bbox.min.y
  cam.height = [mh, mw / (1600.0 / 1200.0)].max * 1.06
  view.camera = cam
end

def pl_iso_camera(model, bbox = nil)
  view = model.active_view
  custom = !bbox.nil?
  bbox ||= model.bounds
  center = bbox.center
  diag = bbox.diagonal
  d = diag * 5.0
  eye = Geom::Point3d.new(center.x + d * 0.5, center.y - d * 0.6, center.z + d * 0.7)
  cam = Sketchup::Camera.new(eye, center, Geom::Vector3d.new(0, 0, 1))
  cam.perspective = false
  cam.height = diag * 1.2
  view.camera = cam
  view.zoom_extents unless custom   # custom bbox = enquadra so o comodo, nao o apê
end

def pl_png(model, path)
  model.active_view.write_image(
    filename: path, width: 1600, height: 1200, antialias: true, transparent: false)
end

def pl_material(model, name, rgb, tex_path = nil, tile = 40)
  m = model.materials[name]
  return m if m
  m = model.materials.add(name)
  m.color = Sketchup::Color.new(rgb[0], rgb[1], rgb[2])
  m.alpha = 1.0
  # A CORRECAO DO BUG (FP-036): o path HUMANO/interativo tambem aplica textura por kind, nao so
  # o V-Ray. Sem png (ou arquivo ausente) -> cor chapada = comportamento anterior (fallback seguro).
  if tex_path && File.exist?(tex_path)
    begin
      m.texture = tex_path
      m.texture.size = [tile, tile] if m.texture   # tamanho fisico do tile em polegadas
    rescue StandardError
      # textura falhou -> fica so na cor solida (nao aborta a peca)
    end
  end
  m
end

def pl_run
  boxes = JSON.parse(ENV['LAYOUT_BOXES'] || '[]')
  model = Sketchup.active_model
  log = []

  # FP-036: mapa de textura por KIND (de style_spec.texture_map_for), injetado pelo Python.
  # kind = FONTE UNICA -> cada ph_<kind> recebe SO a textura do seu kind (nunca a 1a peca em tudo).
  tex_map = (JSON.parse(ENV['LAYOUT_TEX_MAP'] || '{}') rescue {})    # kind -> png
  tile_map = (JSON.parse(ENV['LAYOUT_TILE_MAP'] || '{}') rescue {})  # kind -> tile_in (de tile_map_for)
  tex_dir = ENV['LAYOUT_TEX_DIR']
  tex_logged = {}    # dedup do log por kind (inclui MISS)
  tex_applied = {}   # so os kinds que REALMENTE receberam textura (File.exist? true)

  # BEFORE: shell puro (sem moveis)
  if ENV['LAYOUT_BEFORE']
    pl_top_camera(model)
    pl_png(model, ENV['LAYOUT_BEFORE'])
    log << "BEFORE render -> #{File.basename(ENV['LAYOUT_BEFORE'])}"
  end

  ents = model.active_entities
  # EDITABILIDADE (Felipe 2026-06-18): cada MOVEL = um GRUPO TOP-LEVEL nomeado, pra
  # clique UNICO selecionar SO aquele movel (nao o apê todo). Organizacao por COMODO
  # via TAG/Layer (Outliner mostra; nao aninha a selecao). Nada de mega-grupo 'Mobilia'.
  mod_groups = {}
  furn_bb = Geom::BoundingBox.new
  placed = 0
  boxes.each do |b|
    begin
      h = b['h_in'].to_f
      z0 = (b['z0_in'] || 0).to_f      # base elevada (ex.: armario aereo flutua sobre a bancada)
      room = (b['room'] || 'Apto').to_s
      mod  = (b['module'] || b['kind'] || 'Movel').to_s
      mkey = "#{room}|#{mod}"
      tag = (model.layers[room] || model.layers.add(room))
      mg = mod_groups[mkey]
      if mg.nil?
        mg = ents.add_group            # MOVEL = grupo top-level (selecionavel sozinho)
        mg.name = "#{room} · #{mod}"
        mg.layer = tag
        mod_groups[mkey] = mg
      end
      g = mg.entities.add_group        # a peca, DENTRO do movel
      g.name = b['label'] || b['kind']
      # PERFIL: peca com face 3D no mundo + vetor de extrusao (place_sofa_boxes) ->
      # desenha as CURVAS reais (coroamento/arredondado) em vez de caixa. Padrao p/
      # QUALQUER peca com profile (sofa hoje; qualquer movel refinado no futuro).
      if (pw_face = b['profile_world'])
        ppts = pw_face.map { |c| Geom::Point3d.new(c[0].to_f, c[1].to_f, c[2].to_f) }
        pface = g.entities.add_face(ppts)
        if pface
          ev = b['extrude_vec'] || [0, 0, 0]
          pvec = Geom::Vector3d.new(ev[0].to_f, ev[1].to_f, ev[2].to_f)
          d = pvec.dot(pface.normal)
          pface.pushpull(d.abs < 1e-6 ? pvec.length : d)
        end
        # suaviza as arestas do perfil: sem isto as facetas do coroamento viram
        # 'ranhuras' e o sofa parece faceteado/rugoso, nao curvo liso (regra do harness).
        g.entities.grep(Sketchup::Edge).each do |e|
          if e.faces.length == 2 && e.faces[0].normal.angle_between(e.faces[1].normal) < 0.45
            e.soft = true
            e.smooth = true
          end
        end
      else
      # desenha o POLIGONO real (cantos) na cota z0; almofadas ganham chanfro no topo
      # (Visual Quality Layer: nao parecer cubo/game asset)
      pts = (b['corners'] || []).map { |c| Geom::Point3d.new(c[0].to_f, c[1].to_f, z0) }
      bev = %w[seat_cushion back_cushion arm colchao travesseiro manta].include?(b['kind']) ? (0.04 * 39.3700787402) : 0.0
      face = g.entities.add_face(pts)
      if bev > 0 && h > bev * 1.6
        face.pushpull(face.normal.z >= 0 ? (h - bev) : -(h - bev))
        topz = z0 + h - bev
        top = g.entities.grep(Sketchup::Face).find { |f| f.normal.z.abs > 0.9 && (f.bounds.center.z - topz).abs < 0.3 }
        if top
          bb = top.bounds
          ix0, iy0, ix1, iy1 = bb.min.x + bev, bb.min.y + bev, bb.max.x - bev, bb.max.y - bev
          if ix1 > ix0 && iy1 > iy0
            th = z0 + h
            if b['kind'] == 'arm'   # braco = casca INCLINADA (frustum), sem degrau (GPT)
              bp = [[bb.min.x, bb.min.y, topz], [bb.max.x, bb.min.y, topz], [bb.max.x, bb.max.y, topz], [bb.min.x, bb.max.y, topz]].map { |p| Geom::Point3d.new(*p) }
              tp = [[ix0, iy0, th], [ix1, iy0, th], [ix1, iy1, th], [ix0, iy1, th]].map { |p| Geom::Point3d.new(*p) }
              top.erase!
              4.times { |i| j = (i + 1) % 4; begin; g.entities.add_face(bp[i], bp[j], tp[j], tp[i]); rescue StandardError; end }
              begin; g.entities.add_face(tp); rescue StandardError; end
            else                    # almofada = topo inset levantado (degrau) — GPT aprovou
              ip = [[ix0, iy0, topz], [ix1, iy0, topz], [ix1, iy1, topz], [ix0, iy1, topz]].map { |p| Geom::Point3d.new(*p) }
              g.entities.add_face(ip).pushpull(bev)
            end
          end
        end
      else
        face.pushpull(face.normal.z >= 0 ? h : -h)
      end
      end   # fecha o if (pw_face = b['profile_world'])
      kind = b['kind']
      # FP-037: prefere o material JA RESOLVIDO por (familia,kind) no Python (mat_name/tex_png/
      # tile_in) — assim rack.base=madeira e sofa.base=grafite viram materiais SEPARADOS. Sem os
      # campos (ex. slice sem attach) cai no fallback FP-036 por kind (tex_map[kind]).
      mat_name = b['mat_name'] || "ph_#{kind}"
      png = b.key?('tex_png') ? b['tex_png'] : tex_map[kind]
      tile = (b['tile_in'] || tile_map[kind] || 40).to_f
      tex_path = (png && tex_dir) ? File.join(tex_dir, png) : nil
      if png && tex_dir && !tex_logged[mat_name]                 # log por material (invariante auditavel)
        tex_logged[mat_name] = true
        if File.exist?(tex_path)
          tex_applied[mat_name] = true                           # so conta o que de fato texturizou
          log << "  tex #{mat_name} <- #{png} (tile #{tile.round})"
        else
          log << "  tex MISS #{mat_name}: #{png} ausente -> cor chapada"
        end
      end
      mat = pl_material(model, mat_name, b['rgb'] || [120, 120, 120], tex_path, tile)
      g.material = mat
      placed += 1
      bw = (b['x1'].to_f - b['x0'].to_f).round
      bd = (b['y1'].to_f - b['y0'].to_f).round
      tag = b['ambiguous'] ? ' [TV AMBIGUOUS]' : ''
      log << "  #{b['kind']} bbox #{bw}x#{bd}x#{h.round} in#{tag}"
    rescue StandardError => e
      log << "  FAIL #{b['kind']}: #{e.class}: #{e.message}"
    end
  end
  log << "placed #{placed}/#{boxes.size} placeholders"
  log << "texturas aplicadas: #{tex_applied.size} kind(s) via LAYOUT_TEX_MAP"
  log << "MOVEIS (comodo | movel): #{mod_groups.keys.sort.join(' ; ')}"
  mod_groups.each_value { |mg| (furn_bb.add(mg.bounds) rescue nil) }
  # TRAVA o shell (paredes/piso/portas/janelas) — mover/editar movel NAO atrapalha a base
  nlock = 0
  ents.grep(Sketchup::Group).each do |gp|
    nm = gp.name.to_s
    if %w[PlanShell Floor_Group DoorLeaf Window GlazedBalcony SoftBarrier PassageMarker].any? { |p| nm.start_with?(p) }
      (gp.locked = true; nlock += 1) rescue nil
    end
  end
  log << "shell travado: #{nlock} grupos (paredes/piso/portas)"

  # AFTER: shell + moveis. LAYOUT_ZOOM_GROUP enquadra SO o comodo mobiliado
  # (bounds do grupo de moveis + folga p/ pegar as paredes), nao o apê inteiro.
  zoom_bb = nil
  if ENV['LAYOUT_ZOOM_GROUP'] && placed > 0
    pb = furn_bb
    zoom_bb = Geom::BoundingBox.new
    pad = 48.0   # ~1.2 m de folga
    zoom_bb.add([pb.min.x - pad, pb.min.y - pad, pb.min.z])
    zoom_bb.add([pb.max.x + pad, pb.max.y + pad, pb.max.z + pad])
  end
  if ENV['LAYOUT_AFTER_TOP']
    pl_top_camera(model, zoom_bb)
    pl_png(model, ENV['LAYOUT_AFTER_TOP'])
    log << "AFTER top -> #{File.basename(ENV['LAYOUT_AFTER_TOP'])}"
  end
  if ENV['LAYOUT_AFTER_ISO']
    pl_iso_camera(model, zoom_bb)
    pl_png(model, ENV['LAYOUT_AFTER_ISO'])
    log << "AFTER iso -> #{File.basename(ENV['LAYOUT_AFTER_ISO'])}"
  end

  if ENV['LAYOUT_OUT']
    model.save(ENV['LAYOUT_OUT'])
    log << "saved -> #{File.basename(ENV['LAYOUT_OUT'])}"
  end

  # LOG por ultimo: e o sinal de "done" pro Python
  File.write(ENV['LAYOUT_LOG'] || 'layout_place_log.txt', log.join("\n"))
end

pl_run
